src/middlewared/middlewared/plugins/auth_/authenticate.py

```diff
@@ -0,0 +1,78 @@
+import crypt
+import hmac
+
+from middlewared.service import Service, private
+
+
+class AuthService(Service):
+
+    class Config:
+        cli_namespace = 'auth'
+
+    @private
+    async def authenticate(self, username, password):
+        if '@' in username:
+            if (await self.middleware.call('datastore.config', 'system.settings'))['stg_ds_auth']:
+                return await self.ds_authenticate(username, password)
+            else:
+                return None
+        else:
+            return await self.local_authenticate(username, password)
+
+    @private
+    async def local_authenticate(self, username, password):
+        try:
+            user = await self.middleware.call(
+                'datastore.query',
+                'account.bsdusers',
+                [('bsdusr_username', '=', username)],
+                {'get': True},
+            )
```

Comment:
Shouldn't you be checking for entries that are `locked` or have `password_disabled`?